File: packages/sdk-python/capybara_code/approvals.py
"""Approval decision helpers for interactive / headless clients."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal


DecisionKind = Literal["allow_once", "allow_session", "deny"]


@dataclass(frozen=True, slots=True)
class ApprovalDecision:
    kind: DecisionKind
    reason: str | None = None

    @staticmethod
    def allow_once() -> ApprovalDecision:
        return ApprovalDecision(kind="allow_once")

    @staticmethod
    def allow_session() -> ApprovalDecision:
        return ApprovalDecision(kind="allow_session")

    @staticmethod
    def deny(reason: str) -> ApprovalDecision:
        return ApprovalDecision(kind="deny", reason=reason)

    def to_payload(self) -> dict[str, Any]:
        if self.kind == "deny":
            return {"kind": "deny", "reason": self.reason or "denied"}
        return {"kind": self.kind}


@dataclass(frozen=True, slots=True)
class ApprovalRequest:
    approval_id: str
    session_id: str
    turn_id: str
    title: str
    summary: str
    action_hash: str
    network: bool = False
    raw: dict[str, Any] | None = None

File: packages/sdk-typescript/src/approvals.ts
/**
 * Approval callback contracts for interactive App Protocol sessions.
 */

export interface ApprovalRequest {
  readonly approvalId: string;
  readonly sessionId: string;
  readonly turnId?: string;
  readonly kind: string;
  readonly summary: string;
  readonly risk?: string;
  readonly details?: Readonly<Record<string, unknown>>;
}

export type ApprovalDecision =
  | { readonly decision: "allow" | "allow_once" | "deny"; readonly reason?: string }
  | { readonly decision: "allow_session"; readonly reason?: string };

export type ApprovalHandler = (
  request: ApprovalRequest,
) => ApprovalDecision | Promise<ApprovalDecision>;

export interface ApprovalHooks {
  readonly onApproval?: ApprovalHandler;
}

export async function resolveApproval(
  hooks: ApprovalHooks | undefined,
  request: ApprovalRequest,
): Promise<ApprovalDecision> {
  const handler = hooks?.onApproval;
  if (handler === undefined) {
    return { decision: "deny", reason: "no approval handler registered" };
  }
  return handler(request);
}
